File: app/src/main/java/io/github/fvasco/pinpoi/importer/KmlImporter.java
package io.github.fvasco.pinpoi.importer;

import android.util.Log;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;

/**
 * KML importer
 *
 * @author Francesco Vasco
 */
public class KmlImporter extends AbstractXmlImporter {

    private double latitude, longitude;
    private int coordinateCount;

    @Override
    protected void handleStartTag() {
        switch (tag) {
            case "Placemark":
                newPlacemark();
                latitude = 0;
                longitude = 0;
                coordinateCount = 0;
        }
    }

    @Override
    protected void handleEndTag() throws IOException {
        if (placemark == null) {
            if ("href".equals(tag) && checkCurrentPath("kml", "Document", "NetworkLink", "Url")) {
                final String href = text;
                final AbstractImporter delegateImporter = ImporterFacade.createImporter(href);
                Log.i(KmlImporter.class.getSimpleName(), "NetworkLink href " + href + " importer " + delegateImporter);
                if (delegateImporter != null) {
                    final InputStream inputStream = new URL(href).openStream();
                    try {
                        delegateImporter.configureFrom(this);
                        delegateImporter.importPlacemarks(inputStream);
                    } finally {
                        inputStream.close();
                    }
                }
            }
        } else {
            switch (tag) {
                case "Placemark":
                    if (coordinateCount > 0) {
                        // set placemark to center
                        placemark.setLongitude((float) (latitude / (double) coordinateCount));
                        placemark.setLatitude((float) (longitude / (double) coordinateCount));
                    }
                    importPlacemark();
                    break;
                case "name":
                    placemark.setName(text);
                    break;
                case "description":
                    placemark.setDescription(text);
                    break;
                case "coordinates":
                    // read multiple lines if present (point, line, polygon)
                    for (final String line : text.trim().split("\\s+")) {
                        // format: longitude, latitute, altitude
                        final String[] coordinates = line.split(",", 3);
                        latitude += Double.parseDouble(coordinates[0]);
                        longitude += Double.parseDouble(coordinates[1]);
                        ++coordinateCount;
                    }
                    break;
            }
        }
    }
}

File: app/src/main/java/io/github/fvasco/pinpoi/importer/KmlImporter.kt
package io.github.fvasco.pinpoi.importer

import android.util.Log
import io.github.fvasco.pinpoi.util.Coordinates
import io.github.fvasco.pinpoi.util.makeURL
import java.io.IOException

/**
 * KML importer
 * @author Francesco Vasco
 */
class KmlImporter : AbstractXmlImporter() {

    private var documentNamespace: String? = null
    private var latitude: Float = Float.NaN
    private var longitude: Float = Float.NaN

    override fun handleStartTag() {
        // skip extension
        if (tag == "kml" && documentNamespace == null) documentNamespace = namespace
        if (documentNamespace != namespace) return

        when (tag) {
            "Placemark" -> {
                newPlacemark()
                latitude = Float.NaN
                longitude = Float.NaN
            }
        }
    }

    @Throws(IOException::class)
    override fun handleEndTag() {
        if (documentNamespace != namespace) return

        if (placemark == null) {
            if ("href" == tag && checkCurrentPath("kml", "Document", "NetworkLink", "Url")) {
                val href = text
                Log.e(KmlImporter::class.java.simpleName, "NetworkLink href $href")
                val connection = makeURL(href).openConnection()
                connection.addRequestProperty(
                    "Accept",
                    FileFormatFilter.KML.validMimeTypes.joinToString(postfix = ", */*;q=0.1")
                )
                connection.connect()
                connection.getInputStream().use { inputStream ->
                    val delegateImporter = KmlImporter()
                    delegateImporter.configureFrom(this)
                    delegateImporter.importPlacemarks(inputStream)
                }
            }
        } else {
            val p = placemark ?: return
            when (tag) {
                "Placemark" -> {
                    if (!latitude.isNaN() && !longitude.isNaN()) {
                        // set placemark to center
                        p.coordinates = Coordinates(longitude = longitude, latitude = latitude)
                        importPlacemark()
                    }
                    placemark = null
                }

                "name" -> p.name = text
                "description" -> p.description = text
                "coordinates" -> // read multiple lines if present (point, line, polygon)
                    for (line in text.trim { it <= ' ' }.split("\\s+".toRegex())
                        .dropLastWhile { it.isEmpty() }) {
                        // format: longitude, latitude, altitude
                        val coordinates = line.split(',', limit = 3).takeIf { it.size in 2..3 }
                            ?: continue
                        val lon = coordinates[0].toFloatOrNull() ?: continue
                        val lat = coordinates[1].toFloatOrNull() ?: continue
                        latitude = lat
                        longitude = lon
                        break
                    }
            }
        }
    }
}